examples/bsc-p2p/src/block_import/service.rs

```diff
@@ -0,0 +1,266 @@
+use alloy_primitives::{BlockNumber, B256};
+use alloy_rpc_types::engine::{ForkchoiceState, PayloadStatusEnum};
+use futures::{stream::FuturesUnordered, StreamExt};
+use reth_engine_primitives::{BeaconConsensusEngineHandle, EngineTypes};
+use reth_network::{
+    import::{BlockImportError, BlockImportOutcome, BlockValidation},
+    message::NewBlockMessage,
+};
+use reth_network_api::PeerId;
+use reth_payload_primitives::{BuiltPayload, EngineApiMessageVersion, PayloadTypes};
+use reth_primitives::NodePrimitives;
+use reth_primitives_traits::{AlloyBlockHeader, Block};
+use reth_provider::BlockNumReader;
+use std::{
+    cmp::Ordering,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+
+/// The block type for a given engine
+type BscBlock<T> =
+    <<<T as PayloadTypes>::BuiltPayload as BuiltPayload>::Primitives as NodePrimitives>::Block;
+
+/// Network message containing a new block
+pub(crate) type BlockMsg<T> = NewBlockMessage<BscBlock<T>>;
+
+/// Import outcome for a block
+pub(crate) type Outcome<T> = BlockImportOutcome<BscBlock<T>>;
+
+/// Channel message type for incoming blocks
+type IncomingBlock<T> = (BlockMsg<T>, PeerId);
+
+/// Future that processes a block import and returns its outcome
+type ImportFuture<T> = Pin<Box<dyn Future<Output = Outcome<T>> + Send + Sync>>;
+
+pub struct ImportHandle<T: EngineTypes> {
+    /// Send the new block to the service
+    to_import: UnboundedSender<IncomingBlock<T>>,
+    /// Receive the outcome of the import
+    import_outcome: UnboundedReceiver<Outcome<T>>,
+}
+
+impl<T: EngineTypes> ImportHandle<T> {
+    /// Create a new handle with the provided channels
+    pub fn new(
+        to_import: UnboundedSender<IncomingBlock<T>>,
+        import_outcome: UnboundedReceiver<Outcome<T>>,
+    ) -> Self {
+        Self { to_import, import_outcome }
+    }
+
+    /// Sends the block to import to the service
+    pub fn send_block(
+        &self,
+        block: BlockMsg<T>,
+        peer_id: PeerId,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        self.to_import.send((block, peer_id))?;
```

Comment:
we should use the same error type here and mention that this returns an error if the channel is closed